kolibri/core/auth/constants/demographics.py

```diff
@@ -14,3 +22,148 @@
+            # If we don't do this, every time we initialize, Django thinks it has changed
+            # and will try to create a new migration.
+            "language": {"type": "string", "enum": sorted(KOLIBRI_SUPPORTED_LANGUAGES)},
+            "message": {"type": "string"},
+        },
+    },
+    "optional": True,
+}
+
+
+custom_demographic_field_schema = {
+    "type": "object",
+    "properties": {
+        "id": {
+            "type": "string",
+        },
+        "description": {"type": "string"},
+        "enumValues": {
+            "type": "array",
+            "items": {
+                "type": "object",
+                "properties": {
+                    "value": {"type": "string"},
+                    "defaultLabel": {"type": "string"},
+                    "translations": translations_schema,
+                },
+            },
+        },
+        "translations": translations_schema,
+    },
+}
+
+
+custom_demographics_schema = {
+    "type": "array",
+    "items": custom_demographic_field_schema,
+    "optional": True,
+}
+
+
+@deconstructible
+class UniqueIdsValidator(NoRepeatedValueJSONArrayValidator):
+    def __init__(self, custom_demographics_key):
+        super(UniqueIdsValidator, self).__init__(
+            array_key=custom_demographics_key, object_key="id"
+        )
+
+
+unique_translations_validator = NoRepeatedValueJSONArrayValidator(
+    array_key="translations",
+    object_key="language",
+)
+
+
+@deconstructible
+class DescriptionTranslationValidator(object):
+    def __init__(self, custom_demographics_key):
+        self.custom_demographics_key = custom_demographics_key
+
+    def __call__(self, value):
+        for item in value.get(self.custom_demographics_key, []):
+            try:
+                unique_translations_validator(item)
+            except ValidationError:
+                raise ValidationError(
+                    "User facing description translations for '{} ({})' must be unique by language".format(
+                        item["description"], item["id"]
+                    ),
+                    code="invalid",
+                )
+        return value
+
+
+unique_value_validator = NoRepeatedValueJSONArrayValidator(
+    object_key="value",
+)
+
+
+@deconstructible
+class EnumValuesValidator(object):
+    def __init__(self, custom_demographics_key):
+        self.custom_demographics_key = custom_demographics_key
+
+    def __call__(self, value):
+        for item in value.get(self.custom_demographics_key, []):
+            enum_values = item.get("enumValues", [])
+            try:
+                unique_value_validator(enum_values)
+            except ValidationError:
+                raise ValidationError(
+                    "Possible values for '{} ({})' must be unique".format(
+                        item["description"], item["id"]
+                    ),
+                    code="invalid",
+                )
+        return value
+
+
+@deconstructible
```

Comment:
I looked up what `@deconstructible` does and thought it was weird that the `FacilityUserDemographicValidator` class didn't have the decorator -- but then I looked at `JSON_Schema_Validator` and saw that it was decorated.

So, just to be sure then, class decorators are inherited -- or, are they not inherited and the class below doesn't need the `@deconstructible` decorator at all?

Reply:
I didn't explicitly add it to the `FacilityUserDemographicValidator` because it isn't used in the validation for any field definitions within a model (and so doesn't need to be deconstructed and used within a migration) - it may incidentally be marked as such because of the inheritance though.